### PR TITLE
fix(react-intl): add typescript as peerDependency, fix #2066

### LINKED ISSUE #2066: Not working with create-react-app
**Which package?**
react-intl: "^5.8.0",
node: v12.4.0
npm: v6.14.8

**Describe the bug**
Not working with create-react-app with typescript.

**To Reproduce**

1. `npx create-react-app my-app --template typescript`
2. ` npm i -S react react-intl`

**Expected behavior**
It should be working properly.

**Screenshots**
Screenshot_20200907_200919

**Desktop (please complete the following information):**

- OS:Manajro
- Browser: Chrome
- Version 84.0.4147.125 (Official Build) (64-bit)

### PR DESCRIPTION
**Pull request by longlho**

### CODE REVIEW

File: packages/react-intl/package.json [hunk 1]
```diff
@@ -141,7 +141,8 @@
     "shallow-equal": "^1.2.1"
   },
   "peerDependencies": {
-    "react": "^16.3.0"
+    "react": "^16.3.0",
+    "typescript": "4"
```
Comment (pyrocat101) [inline packages/react-intl/package.json]: I think we just need 3.8. So maybe `^3.8 || 4`?

**Comment by chrisbobbe** [comment]
Hmm, this might have been better released as a breaking change, instead of in the patch release [5.8.1](https://github.com/formatjs/formatjs/releases/tag/react-intl%405.8.1). Having routinely bumped up the version of `react-intl` to its latest 5.x, I'm suddenly required to install TypeScript as a direct dependency, even though I don't use TypeScript at all (I use Flow).

`peerDependencies` are generally considered to be part of a package's public API: https://github.com/semver/semver/issues/502#issuecomment-466501843.

**Comment by longlho** [comment]
Hmm why are you required to?

**Comment by chrisbobbe** [comment]
I'm required to add `typescript` to my own project's `dependencies` or `devDependencies` because `typescript` was added under `peerDependencies` in this PR.

Additionally, when I do add it, I'm required to use a version within the specific range of "3.8 || 4". This is currently impossible for me, because another package I'm using lists `typescript` under its `peerDependencies`, but with a range that doesn't intersect with that range.

Concretely, after upgrading `react-intl`, I get this warning when I remove `node_modules` and run `yarn`:

```
warning " > react-intl@5.8.6" has incorrect peer dependency "typescript@3.8 || 4".
```

**Comment by longlho** [comment]
Yeah you can ignore the warning

**Comment by chrisbobbe** [comment]
> Additionally, when I do add it, I'm required to use a version within the specific range of "3.8 || 4". This is currently impossible for me, because another package I'm using lists `typescript` under its `peerDependencies`, but with a range that doesn't intersect with that range.

This isn't necessarily `react-intl`'s fault—I generally assume there's a valid reason for a specific version range to be given for a peer dependency. So I'm currently analyzing whether it's better for my project (for the time being) to stay on `react-intl` 5.8.0, or use a different version of that other package. But I would more expect to have to do this exercise as part of a major-version upgrade of `react-intl`, not a patch- or minor-version upgrade.

**Comment by chrisbobbe** [comment]
> Yeah you can ignore the warning

I'd prefer not to. The warning is there for a reason: `react-intl` has made an assertion about what my project's dependencies must include, and respecting that assertion is the natural thing to do.

We've found our project to be much more stable when we don't just ignore or paper over warnings, including peer-dependency warnings. One of our highest priorities is making it easy for new contributors to get started—and letting lots of warnings accumulate in a routine setup step like `yarn` just adds unnecessary confusion.

**Comment by longlho** [comment]
Up to you

**Comment by longlho** [comment]
Or rather let's put it this way: do you have an alternative solution? 

Bumping major version doesn't solve this problem.

**Comment by chrisbobbe** [comment]
> Bumping major version doesn't solve this problem.

I appreciate it when packages wait for a major-version release to make a public API change. Is this where you disagree? Or would you say that introducing peer-dependency constraints, or dropping previously valid versions from a peer dependency's version range, are not public API changes? [Above](https://github.com/formatjs/formatjs/pull/2067#issuecomment-714079989), I linked to a discussion of that question: https://github.com/semver/semver/issues/502#issuecomment-466501843

**Comment by longlho** [comment]
Those are not public API changes. We don't bump major when we drop node 8, or react 16.3, or old CLDR dataset

**Comment by chrisbobbe** [comment]
> Those are not public API changes. We don't bump major when we drop node 8, or react 16.3, or old CLDR dataset

Since you're reasoning from `react-intl`'s habits, and not from principles (or even from other projects' habits), I understand your argument not as "these are not public API changes", but as "these are not public API changes *for `react-intl`*"—that is, `react-intl`'s particular policy is to make this kind of breaking change (and some other kinds, apparently) at random. I'll keep a mental note of it; I'm always careful when I take dependencies' new versions, but I'll start treating `react-intl`'s with extra skepticism and distrust.

**Comment by longlho** [comment]
Yes you should always pin to a specific version instead of taking ranges and run you test suite when you upgrade.

**Pull request merged**